dpcpp/distributed/partition_kernels.dp.cpp

```diff
@@ -0,0 +1,108 @@
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/distributed/partition_kernels.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace dpcpp {
+namespace partition {
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+void build_starting_indices(std::shared_ptr<const DefaultExecutor> exec,
+                            const GlobalIndexType* range_offsets,
+                            const int* range_parts, size_type num_ranges,
+                            int num_parts, int& num_empty_parts,
+                            LocalIndexType* ranks, LocalIndexType* sizes)
+{
+    Array<LocalIndexType> range_sizes{exec, num_ranges};
+    // num_parts sentinel at the end
+    Array<comm_index_type> tmp_part_ids{exec, num_ranges + 1};
+    Array<GlobalIndexType> permutation{exec, num_ranges};
+    // set sizes to 0 in case of empty parts
+    components::fill_array(exec, sizes, num_parts, LocalIndexType{});
+
+    run_kernel(
+        exec,
+        [] GKO_KERNEL(auto i, auto num_ranges, auto num_parts,
+                      auto range_offsets, auto range_parts, auto range_sizes,
+                      auto tmp_part_ids, auto permutation) {
+            if (i == 0) {
+                // set sentinel value at the end
+                tmp_part_ids[num_ranges] = num_parts;
+            }
+            range_sizes[i] = range_offsets[i + 1] - range_offsets[i];
+            tmp_part_ids[i] = range_parts[i];
+            permutation[i] = static_cast<GlobalIndexType>(i);
+        },
+        num_ranges, num_ranges, num_parts, range_offsets, range_parts,
+        range_sizes, tmp_part_ids, permutation);
+
+    // group sizes by part ID
+    // TODO oneDPL has stable_sort and views::zip
+    // compute inclusive prefix sum for each part
+    // TODO compute "row_ptrs" for tmp_part_ids
+    // TODO compute prefix_sum over range_sizes
+    // TODO compute adjacent differences, set -part_size at part boundaries
+    // TODO compute prefix_sum again
+    // write back the results
+    // TODO this needs to be adapted to the output of the algorithm above
+    // TODO count number of zeros in size and store in num_empty_parts
+    run_kernel(
+        exec,
+        [] GKO_KERNEL(auto i, auto grouped_range_ranks, auto grouped_part_ids,
```

Comment:
Since this code is not particularly trivial, I suggest you remove this kernel and put it in a common header file; perhaps `common/unified/partition_kernels.hpp`? Or any other place you think is appropriate. This could apply to the kernel above as well.